### PR TITLE
CORDA-1833: Create a picocli base class

### PR DESCRIPTION
**Pull request by anthonykeenan**
* Merged the shell extension functionality with the CLIUtils class
* Refactored the CLIUtils to be opinionated about as many options as is reasonable in order to provide a basis for a standard UX for Corda CLI applications
* Moved the utils out of the node module into its own module so consumers don't have to depend on node
* Refactored the bootstrapper to use the new functionality

Link to CLI UX guidelines PR: https://github.com/corda/corda/pull/3686

### CODE REVIEW

**Comment by mikehearn** [comment]
Cool, can we have a page in the docsite or code style guide explaining how to use the standard infrastructure?Under `Participate > Contributing`?

**Comment by anthonykeenan** [comment]
And here's one I made earlier: https://github.com/corda/corda/pull/3686

We've left the PR open so we can modify it as we get through the stories. I'll update it for info on the base class once the PR is approved.

**Comment by mikehearn** [comment]
Sweet.

File: tools/cliutils/src/main/kotlin/net/corda/cliutils/CordaVersionProvider.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import com.jcabi.manifests.Manifests
+import picocli.CommandLine
+
+/**
+ * Simple version printing when command is called with --version or -V flag. Assuming that we reuse Corda-Release-Version and Corda-Revision
+ * in the manifest file.
+ */
+class CordaVersionProvider : CommandLine.IVersionProvider {
+    companion object {
+        val releaseVersion: String by lazy { Manifests.read("Corda-Release-Version") }
```
Comment (tlil) [inline tools/cliutils/src/main/kotlin/net/corda/cliutils/CordaVersionProvider.kt]: Reuse this in `getVersion()`? Or why is this duplicated?

File: tools/cliutils/src/main/kotlin/net/corda/cliutils/ConfigFilePathArgsParser.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import net.corda.core.internal.exists
+import net.corda.core.internal.isReadable
+import picocli.CommandLine
+import java.nio.file.Path
+
+/**
+ * Something heavily used in network services, I am not sure it's of much use in corda, but who knows. Definitely it was the key to making DevOps happy.
+ * Add it as
+ * `@CommandLine.Mixin
```
Comment (tlil) [inline tools/cliutils/src/main/kotlin/net/corda/cliutils/ConfigFilePathArgsParser.kt]: Update kdoc comment? :)

File: tools/cliutils/src/main/kotlin/net/corda/cliutils/CordaCliWrapper.kt [hunk 3]
```diff
@@ -75,6 +39,26 @@ interface Validated {
...
+                DefaultExceptionHandler<List<Any>>().useErr(System.err).useAnsi(Help.Ansi.AUTO),
+                *args)
+    } catch (e: ExecutionException) {
+        val throwable = e.cause ?: e
+        if (this.verbose) {
+            throwable.printStackTrace()
+        } else {
+            System.err.println("*ERROR*: ${throwable.rootMessage ?: "Use --verbose for more details"}")
+        }
+        exitProcess(1)
```
Comment (tlil) [inline tools/cliutils/src/main/kotlin/net/corda/cliutils/CordaCliWrapper.kt]: As a follow-up, we should introduce customisable exit codes as well; see CORDA-1757

Comment (anthonykeenan) [reply tools/cliutils/src/main/kotlin/net/corda/cliutils/CordaCliWrapper.kt]: Yes, the network bootstrapper doesn't really have any, and I know some work has been done in the node startup, so I've just left this as is for now.

File: tools/cliutils/src/main/kotlin/net/corda/cliutils/ConfigFilePathArgsParser.kt (hunk 2 above)
Comment (anthonykeenan) [reply tools/cliutils/src/main/kotlin/net/corda/cliutils/ConfigFilePathArgsParser.kt]: 👍

File: tools/cliutils/src/main/kotlin/net/corda/cliutils/CordaVersionProvider.kt (hunk 1 above)
Comment (anthonykeenan) [reply tools/cliutils/src/main/kotlin/net/corda/cliutils/CordaVersionProvider.kt]: 👍 - I was sure I'd done that.

**Comment by anthonykeenan** [comment]
I also fixed a bug that was running the program twice.

**Comment by anthonykeenan** [comment]
Yes, I tested it manually apart from actually bootstrapping a network, but I've just done that now and it works fine.

**Comment by tlil** [comment]
Thanks :+1:

**Pull request merged**